File: kotlin/src/main/kotlin/SnapkitImageURL.kt
package dev.snapkit.imageurl

import android.net.Uri

/**
 * Image transformation parameters
 */
data class TransformOptions(
    /** Image width (pixels) */
    val w: Int? = null,
    /** Image height (pixels) */
    val h: Int? = null,
    /** Resize method */
    val fit: Fit? = null,
    /** Output format */
    val format: Format? = null,
    /** Rotation angle (degrees) */
    val rotation: Int? = null,
    /** Blur strength (0.3-1000) */
    val blur: Int? = null,
    /** Whether to convert to grayscale */
    val grayscale: Boolean? = null,
    /** Whether to flip vertically */
    val flip: Boolean? = null,
    /** Whether to flip horizontally */
    val flop: Boolean? = null,
    /** Region extraction */
    val extract: Extract? = null,
    /** Device Pixel Ratio (1.0-4.0) */
    val dpr: Double? = null,
    /** Image quality (1-100) */
    val quality: Int? = null
) {
    /** Resize method */
    enum class Fit {
        CONTAIN, COVER, FILL, INSIDE, OUTSIDE;

        override fun toString() = name.lowercase()
    }

    /** Output format */
    enum class Format {
        JPEG, PNG, WEBP, AVIF;

        override fun toString() = name.lowercase()
    }

    /** Region extraction */
    data class Extract(
        val x: Int,
        val y: Int,
        val width: Int,
        val height: Int
    )
}

/**
 * Snapkit image URL builder
 *
 * @property organizationName Organization name (used as Snapkit subdomain)
 *
 * @example
 * ```kotlin
 * val builder = SnapkitImageURL("my-org")
 * val imageUrl = builder.build(
 *     url = "https://cdn.cloudfront.net/image.jpg",
 *     transform = TransformOptions(
 *         w = 300,
 *         h = 200,
 *         fit = TransformOptions.Fit.COVER,
 *         format = TransformOptions.Format.WEBP
 *     )
 * )
 * ```
 */
class SnapkitImageURL(private val organizationName: String) {

    /**
     * Generate Snapkit image proxy URL
     *
     * @param url Original image URL
     * @param transform Image transformation options
     * @return Complete image proxy URL
     */
    fun build(url: String, transform: TransformOptions? = null): String {
        val baseUrl = "https://$organizationName.snapkit.dev/image"

        val uriBuilder = Uri.parse(baseUrl).buildUpon()
        uriBuilder.appendQueryParameter("url", url)

        transform?.let {
            val transformString = buildTransformString(it)
            if (transformString.isNotEmpty()) {
                uriBuilder.appendQueryParameter("transform", transformString)
            }
        }

        return uriBuilder.build().toString()
    }

    private fun buildTransformString(options: TransformOptions): String {
        val parts = mutableListOf<String>()

        // Numeric/string value parameters
        options.w?.let { parts.add("w:$it") }
        options.h?.let { parts.add("h:$it") }
        options.fit?.let { parts.add("fit:$it") }
        options.format?.let { parts.add("format:$it") }
        options.rotation?.let { parts.add("rotation:$it") }
        options.blur?.let { parts.add("blur:$it") }
        options.dpr?.let { parts.add("dpr:$it") }
        options.quality?.let { parts.add("quality:$it") }

        // Boolean parameters
        if (options.grayscale == true) parts.add("grayscale")
        if (options.flip == true) parts.add("flip")
        if (options.flop == true) parts.add("flop")

        // extract parameter
        options.extract?.let {
            parts.add("extract:${it.x}-${it.y}-${it.width}-${it.height}")
        }

        return parts.joinToString(",")
    }
}

File: php/src/SnapkitImageURL.php
<?php

namespace Snapkit\ImageURL;

/**
 * Image transformation parameters
 */
class TransformOptions
{
    /** @var int|null Image width (pixels) */
    public ?int $w = null;

    /** @var int|null Image height (pixels) */
    public ?int $h = null;

    /** @var string|null Resize mode (contain, cover, fill, inside, outside) */
    public ?string $fit = null;

    /** @var string|null Output format (jpeg, png, webp, avif) */
    public ?string $format = null;

    /** @var int|null Rotation angle (degrees) */
    public ?int $rotation = null;

    /** @var int|null Blur intensity (0.3-1000) */
    public ?int $blur = null;

    /** @var bool|null Whether to convert to grayscale */
    public ?bool $grayscale = null;

    /** @var bool|null Whether to flip vertically */
    public ?bool $flip = null;

    /** @var bool|null Whether to flip horizontally */
    public ?bool $flop = null;

    /** @var Extract|null Area extraction */
    public ?Extract $extract = null;

    /** @var float|null Device Pixel Ratio (1.0-4.0) */
    public ?float $dpr = null;

    /** @var int|null Image quality (1-100) */
    public ?int $quality = null;

    public function __construct(array $options = [])
    {
        foreach ($options as $key => $value) {
            if (property_exists($this, $key)) {
                $this->$key = $value;
            }
        }
    }
}

/**
 * Area extraction
 */
class Extract
{
    public int $x;
    public int $y;
    public int $width;
    public int $height;

    public function __construct(int $x, int $y, int $width, int $height)
    {
        $this->x = $x;
        $this->y = $y;
        $this->width = $width;
        $this->height = $height;
    }
}

/**
 * Snapkit image URL builder
 *
 * @example
 * ```php
 * $builder = new SnapkitImageURL('my-org');
 * $imageUrl = $builder->build(
 *     'https://cdn.cloudfront.net/image.jpg',
 *     new TransformOptions([
 *         'w' => 300,
 *         'h' => 200,
 *         'fit' => 'cover',
 *         'format' => 'webp'
 *     ])
 * );
 * ```
 */
class SnapkitImageURL
{
    private string $organizationName;

    public function __construct(string $organizationName)
    {
        $this->organizationName = $organizationName;
    }

    /**
     * Generate Snapkit image proxy URL
     *
     * @param string $url Original image URL
     * @param TransformOptions|null $transform Image transformation options
     * @return string Complete image proxy URL
     */
    public function build(string $url, ?TransformOptions $transform = null): string
    {
        $baseUrl = "https://{$this->organizationName}.snapkit.dev/image";

        $queryParams = [
            'url' => $url,
        ];

        if ($transform !== null) {
            $transformString = $this->buildTransformString($transform);
            if (!empty($transformString)) {
                $queryParams['transform'] = $transformString;
            }
        }

        return $baseUrl . '?' . http_build_query($queryParams);
    }

    private function buildTransformString(TransformOptions $options): string
    {
        $parts = [];

        // Numeric/string value parameters
        if ($options->w !== null) {
            $parts[] = "w:{$options->w}";
        }
        if ($options->h !== null) {
            $parts[] = "h:{$options->h}";
        }
        if ($options->fit !== null) {
            $parts[] = "fit:{$options->fit}";
        }
        if ($options->format !== null) {
            $parts[] = "format:{$options->format}";
        }
        if ($options->rotation !== null) {
            $parts[] = "rotation:{$options->rotation}";
        }
        if ($options->blur !== null) {
            $parts[] = "blur:{$options->blur}";
        }
        if ($options->dpr !== null) {
            $parts[] = "dpr:{$options->dpr}";
        }
        if ($options->quality !== null) {
            $parts[] = "quality:{$options->quality}";
        }

        // Boolean parameters
        if ($options->grayscale === true) {
            $parts[] = 'grayscale';
        }
        if ($options->flip === true) {
            $parts[] = 'flip';
        }
        if ($options->flop === true) {
            $parts[] = 'flop';
        }

        // extract parameter
        if ($options->extract !== null) {
            $e = $options->extract;
            $parts[] = "extract:{$e->x}-{$e->y}-{$e->width}-{$e->height}";
        }

        return implode(',', $parts);
    }
}
